### PR TITLE
Fixes #2014: Adds search service to the front-end

### LINKED ISSUE #2014: Have search service work in staging and production
Once #2106 lands, we'll need to setup the staging and production docker files to use this new service as well as ensure it's used in the frontend.

### PR DESCRIPTION
**Pull request by HyperTHD**
## Issue This PR Addresses
This PR closes #2014 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

Now that the search service is functional and merged, we can finally work towards moving it to the front-end. This PR addresses that by adding the search service related arguments to the Dockerfile, the production.yml file, the next.config.js file, the config.ts file, and finally the relevant components that make requests to this endpoint. One bug in the search service is preventing this from working and so I'll need to address that before this PR is ready for review.

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: docker/production.yml [hunk 1]
```diff
@@ -141,6 +143,7 @@ services:
     environment:
       - ELASTIC_URL=http://elasticsearch
       - ELASTIC_PORT=9200
+      - SEARCH_PORT
```
Comment (humphd) [inline docker/production.yml:146]: This shouldn't be necessary, we do it in docker-compose.yml.

**Review by humphd — CHANGES_REQUESTED** [review]
I updated Vercel to add the Search URL env var:

```

14:44:10.365 | Running "npm run build"
-- | --
14:44:10.659 | > @senecacdot/telescope-frontend@3.0.0 build /vercel/workpath0
14:44:10.659 | > next build && next export
14:44:11.683 | info  - Using webpack 5. Reason: future.webpack5 option enabled https://nextjs.org/docs/messages/webpack5
14:44:11.881 | Trying to load missing ENV variables from /.env
14:44:11.881 | Trying to load missing ENV variables from /config/env.development
14:44:11.881 | Using NEXT_PUBLIC_WEB_URL=https://telescope-c64dfl18s-humphd.vercel.app
14:44:11.881 | Using NEXT_PUBLIC_API_URL=https://dev.telescope.cdot.systems
14:44:11.881 | Using NEXT_PUBLIC_IMAGE_URL=https://dev.api.telescope.cdot.systems/v1/image
14:44:11.881 | Using NEXT_PUBLIC_POSTS_URL=https://dev.api.telescope.cdot.systems/v1/posts
14:44:11.881 | Using NEXT_PUBLIC_AUTH_URL=https://dev.api.telescope.cdot.systems/v1/auth
14:44:11.881 | Using NEXT_PUBLIC_SEARCH_URL=https://dev.api.telescope.cdot.systems/v1/search
```

But it's getting a 404 from https://dev.api.telescope.cdot.systems/v1/search/query?text=react&filter=post&page=0.  What am I missing?  Is it the extra `/query`? in that URL?

**Comment by HyperTHD** [comment]
> I updated Vercel to add the Search URL env var:
> 
> ```
> 
> 14:44:10.365 | Running "npm run build"
> -- | --
> 14:44:10.659 | > @senecacdot/telescope-frontend@3.0.0 build /vercel/workpath0
> 14:44:10.659 | > next build && next export
> 14:44:11.683 | info  - Using webpack 5. Reason: future.webpack5 option enabled https://nextjs.org/docs/messages/webpack5
> 14:44:11.881 | Trying to load missing ENV variables from /.env
> 14:44:11.881 | Trying to load missing ENV variables from /config/env.development
> 14:44:11.881 | Using NEXT_PUBLIC_WEB_URL=https://telescope-c64dfl18s-humphd.vercel.app
> 14:44:11.881 | Using NEXT_PUBLIC_API_URL=https://dev.telescope.cdot.systems
> 14:44:11.881 | Using NEXT_PUBLIC_IMAGE_URL=https://dev.api.telescope.cdot.systems/v1/image
> 14:44:11.881 | Using NEXT_PUBLIC_POSTS_URL=https://dev.api.telescope.cdot.systems/v1/posts
> 14:44:11.881 | Using NEXT_PUBLIC_AUTH_URL=https://dev.api.telescope.cdot.systems/v1/auth
> 14:44:11.881 | Using NEXT_PUBLIC_SEARCH_URL=https://dev.api.telescope.cdot.systems/v1/search
> ```
> 
> But it's getting a 404 from https://dev.api.telescope.cdot.systems/v1/search/query?text=react&filter=post&page=0. What am I missing? Is it the extra `/query`? in that URL?

Think it's this piece that @manekenpix mentioned when fixing the search url awhile back

image

File: src/web/src/components/SearchResults.tsx [hunk 2]
```diff
@@ -54,7 +54,9 @@ const SearchResults = () => {
   const { textParam, filter } = useSearchValue();
 
   const prepareUrl = (index: number) =>
-    `${telescopeUrl}/query?text=${encodeURIComponent(textParam)}&filter=${filter}&page=${index}`;
+    `${searchServiceUrl}/query?text=${encodeURIComponent(
```
Comment (manekenpix) [inline src/web/src/components/SearchResults.tsx:57]: it works for me without `query` in the URL

```suggestion
    `${searchServiceUrl}?text=${encodeURIComponent(
```

Comment (manekenpix) [reply src/web/src/components/SearchResults.tsx:57]: If you want to keep using `query`, [this line](https://github.com/Seneca-CDOT/telescope/blob/bf293f64cc7919c832f4f57326e107f916951660/src/api/search/src/index.js#L7) should be:

```js
service.router.use('/query', query);
```

Comment (humphd) [reply src/web/src/components/SearchResults.tsx:57]: We don't want `/query` any more, please remove.

**Review by humphd — CHANGES_REQUESTED** [review]
Looks good except for the `/query` path, please remove that and I'll retest/review.

File: src/api/search/src/routes/query.js [hunk 3]
```diff
@@ -4,12 +4,12 @@ const search = require('../search');
...
 const router = Router();
 
-router.get('/', validateQuery, async (req, res) => {
+router.get('/', validateQuery, async (req, res, next) => {
   try {
     const { text, filter, page, perPage } = req.query;
     res.send(await search(text, filter, page, perPage));
   } catch (error) {
-    next(err);
+    next(error);
```
Comment (humphd) [inline src/api/search/src/routes/query.js:12]: Great catch!

File: src/api/search/src/routes/query.js [hunk 4]
```diff
@@ -4,12 +4,12 @@ const search = require('../search');
 
 const router = Router();
 
-router.get('/', validateQuery, async (req, res) => {
+router.get('/', validateQuery, async (req, res, next) => {
```
Comment (humphd) [inline src/api/search/src/routes/query.js:7]: Excellent catch

**Review by humphd — APPROVED** [review]

**Review by manekenpix — APPROVED** [review]

**Pull request merged**